File: src/main/java/com/HOIIVUtils/clauzewitz/script/AbstractPDX.java
package com.HOIIVUtils.clauzewitz.script;

import com.HOIIVUtils.clausewitz_parser.Node;
import com.HOIIVUtils.clausewitz_parser.NodeValue;
import com.HOIIVUtils.clausewitz_parser.Parser;
import com.HOIIVUtils.clausewitz_parser.ParserException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.util.List;

/**
 * Any object that can be converted to a PDX script block, such as a focus, national focus tree,
 * or event.
 * <p>
 */
public abstract class AbstractPDX<T> implements PDXScript<T> {
    protected T obj;
    protected final List<String> pdxIdentifiers;
    int activeIdentifier = 0;

    public AbstractPDX(String pdxIdentifiers) {
        this.pdxIdentifiers = List.of(pdxIdentifiers);
    }

    public AbstractPDX(String... PDXIdentifiers) {
        this.pdxIdentifiers = List.of(PDXIdentifiers);
    }

    public AbstractPDX(List<String> pdxIdentifiers) {
        this.pdxIdentifiers = pdxIdentifiers;
    }

    protected void usingIdentifier(Node exp) throws UnexpectedIdentifierException {
        for (int i = 0; i < pdxIdentifiers.size(); i++) {
            if (exp.nameEquals(pdxIdentifiers.get(i))) {
                activeIdentifier = i;
                return;
            }
        }
        throw new UnexpectedIdentifierException(exp);
    }

    public void set(T obj) {
        this.obj = obj;
    }

    @SuppressWarnings("unchecked")
    public void set(Node expression) throws UnexpectedIdentifierException, NodeValueTypeException {
        usingIdentifier(expression);
        NodeValue value = expression.value();

        try {
            obj = (T) value.valueObject();
        } catch (ClassCastException e) {
            throw new NodeValueTypeException(expression, e);
        }
    }

    public @Nullable T get() {
        return obj;
    }

    public void loadPDX(Node expression) throws UnexpectedIdentifierException {
        if (expression.name() == null) {
            System.out.println("Error loading PDX script: " + expression);
            return;
        }

        try {
            set(expression);
        } catch (UnexpectedIdentifierException | NodeValueTypeException e) {
            System.out.println("Error loading PDX script:" + e.getMessage() + "\n\t" + expression);
        }
    }

    public void loadPDX(List<Node> expressions) {
        if (expressions == null) return;
        expressions.stream().filter(this::isValidIdentifier)
                .findFirst()
                .ifPresentOrElse(expression -> {
                    try {
                        loadPDX(expression);
                    } catch (UnexpectedIdentifierException e) {
                        throw new RuntimeException(e);
                    }
                }, this::setNull);
    }

    protected void loadPDX(@NotNull File file) {
        if (!file.exists()) {
            System.err.println("Focus tree file does not exist: " + file);
            return;
        }

        /* parser */
        var pdxParser = new Parser(file);
        Node rootNode;
        try {
            rootNode = pdxParser.parse();
        } catch (ParserException e) {
            System.err.println("Error parsing focus tree file: " + file);
            return;
        }
        try {
            loadPDX(rootNode);
        } catch (UnexpectedIdentifierException e) {
            throw new RuntimeException(e);
        }
    }

    public boolean isValidIdentifier(Node node) {
        for (String identifier : pdxIdentifiers) {
            if (node.name().equals(identifier)) {
                return true;
            }
        }
        return false;
    }

    public void setNull() {
        obj = null;
    }

    public void loadOrElse(Node exp, T value) {
        try {
            loadPDX(exp);
        } catch (UnexpectedIdentifierException e) {
            throw new RuntimeException(e);
        }
        if (obj == null) {
            obj = value;
        }
    }

    public @Nullable String toScript() {
        if (obj == null) return null;
        return pdxIdentifiers.get(activeIdentifier) + " = " + obj + "\n";
    }

    public boolean objEquals(AbstractPDX<?> other) {
        if (obj == null) {
            return false;
        }
        if (other.obj == null) {
            return false;
        }
        return obj.equals(other.obj);
    }

    public T getOrElse(T elseValue) {
        return isUndefined() ? elseValue : obj;
    }

    @Override
    public String toString() {
        if (obj == null) {
            return super.toString();
        }
        return obj.toString();
    }

    public boolean isUndefined() {
        return obj == null;
    }

    public String getPDXIdentifier() {
        return pdxIdentifiers.get(activeIdentifier);
    }
}

File: src/main/scala/com/hoi4utils/clausewitz/script/AbstractPDX.scala
package com.hoi4utils.clausewitz.script

import com.hoi4utils.clausewitz.HOIIVUtils
import com.hoi4utils.clausewitz.exceptions.{NodeValueTypeException, UnexpectedIdentifierException}
import com.hoi4utils.clausewitz_parser.{Node, NodeValue, Parser, ParserException}
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.jetbrains.annotations.{NotNull, Nullable}

import java.io.File
import scala.annotation.targetName
import scala.collection.mutable.ListBuffer
import scala.reflect.ClassTag

/**
 * Any object that can be converted to a PDX script block, such as a focus, national focus tree,
 * or event.
 * PDX = Paradox Interactive Clauswitz Engine Modding/Scripting Language
 * <p>
 */
trait AbstractPDX[T](protected val pdxIdentifiers: List[String]) extends PDXScript[T] {
  val LOGGER: Logger = LogManager.getLogger(classOf[AbstractPDX[T]])
  
  private[script] var activeIdentifier = 0
  protected[script] var node: Option[Node] = None

  /**
   * Sets the active identifier to match the given expression, 
   * if it is a valid identifier. Otherwise, throws exception. 
   * @param expr the expression to check, and set the active identifier to the identifier of the expression
   * @throws UnexpectedIdentifierException if the expression is not a valid identifier
   */
  @throws[UnexpectedIdentifierException]
  protected def usingIdentifier(expr: Node): Unit = {
    if (pdxIdentifiers.isEmpty) {
      // all good? 
    }
    else if (pdxIdentifiers.indexWhere(expr.nameEquals) == -1) {
      LOGGER.error("Unexpected identifier: " + expr.name)
      throw new UnexpectedIdentifierException(expr)
    } 
  }

  /**
   * @inheritdoc
   */
  override def setNode(value: T | String | Int | Double | Boolean | ListBuffer[Node] | Null): Unit = {
  // todo?
    if (node.isEmpty) {
      return
    }
    if (value == null) {
      setNull()
      return
    }
    value.match {
      case s: String => node.get.setValue(s)
      case i: Int => node.get.setValue(i)
      case d: Double => node.get.setValue(d)
      case b: Boolean => node.get.setValue(b)
      case _ => throw new RuntimeException("Unsupported type")
    }
  }

  /**
   * @inheritdoc
   */
  @throws[UnexpectedIdentifierException]
  override def set(expression: Node): Unit = {
    usingIdentifier(expression)
    setNode(expression.$)
  }

  /**
   * @inheritdoc
   */
  override def value: Option[T] = {
    node.getOrElse(return None).$ match {
      case value: T => Some(value)
      case _ => None
    }
  }

  /**
   * @inheritdoc
   */
  override def getNode: Option[Node] = node

  /**
   * @inheritdoc
   */
  @throws[UnexpectedIdentifierException]
  override def loadPDX(expression: Node): Unit = {
    if (expression.identifier.isEmpty) {
      if (pdxIdentifiers.nonEmpty || expression.isEmpty) {
        System.out.println("Error loading PDX script: " + expression)
        return
      }
    }
    try set(expression)
    catch {
      case e@(_: UnexpectedIdentifierException | _: NodeValueTypeException) =>
        System.out.println("Error loading PDX script:" + e.getMessage + "\n\t" + expression)
    }
  }

  def loadPDX(expressions: Iterable[Node]): Unit = {
    Option(expressions).foreach { exprs =>
      exprs.find(isValidIdentifier) match {
        case Some(expression) =>
          try loadPDX(expression)
          catch {
            case e: UnexpectedIdentifierException =>
              throw new RuntimeException(e)
          }
        case None => setNull()
      }
    }
  }
  
  protected def loadPDX(file: File): Unit = {
    if (!file.exists) {
      LOGGER.error(s"Focus tree file does not exist: $file")
      return
    }

    try {
      val pdxParser = new Parser(file)
      val rootNode = pdxParser.parse
      loadPDX(rootNode)
    } catch {
      case e: ParserException =>
        LOGGER.error(s"Error parsing focus tree file: $file\n\t${e.getMessage}")
      case e: UnexpectedIdentifierException => throw new RuntimeException(e)
    }
  }

  /**
   * @inheritdoc
   */
  override def isValidIdentifier(node: Node): Boolean = {
    pdxIdentifiers.contains(node.name)
  }
  
  override def isValidID(identifier: String): Boolean = {
    pdxIdentifiers.contains(identifier)
  }

  /**
   * @inheritdoc
   */
  override def clearNode(): Unit = {
    node = None
  }

  /**
   * @inheritdoc
   */
  override def setNull(): Unit = {
    node.foreach(_.setNull())
  }

  override def loadOrElse(exp: Node, value: T): Unit = {
    try loadPDX(exp)
    catch {
      case e: UnexpectedIdentifierException =>
        throw new RuntimeException(e)
    }
    if (node.get.valueIsNull) set(value)
  }

  override def toScript: String = {
    if (node.isEmpty || node.get.isEmpty) return null
    node.get.toScript
  }

  override def equals(other: PDXScript[?]): Boolean = {
    other match {
      case pdx: AbstractPDX[?] =>
        if (node == null) return false
        if (pdx.node == null) return false
        node.equals(pdx.node)
      case _ => false
    }
  }

  /**
   * @inheritdoc
   */
  override def getOrElse(elseValue: T): T = {
    val value = node.getOrElse(return elseValue).value
    value match
      case Some(t: T) => t
      case _ => elseValue
  }

  override def toString: String = {
    if (node.isEmpty || node.get.isEmpty) return super.toString
    node.toString
  }

  override def isUndefined: Boolean = node.isEmpty || node.get.valueIsNull

  override def isDefined: Boolean = !isUndefined

  override def pdxIdentifier: String = {
    if (pdxIdentifiers.isEmpty) return null
    else pdxIdentifiers(activeIdentifier)
  }

  /**
   * Returns true if the value of the node is an instance of the specified class.
   * The implicit class tag is necessary to get around type erasure (preserve A's class at runtime).
   * @param ct
   * @tparam A
   * @return
   */
  def valueIsInstanceOf[A](implicit ct: ClassTag[A]): Boolean = {
    if (node.isEmpty) false
    else ct.runtimeClass.isInstance(node.get.$)
  }

  def getPDXTypeName: String = {
    pdxIdentifier
  }

  def schema(): PDXSchema[T] = {
    var schema = new PDXSchema[T](pdxIdentifiers*)
    null.asInstanceOf[PDXSchema[T]]  // todo no
  }

}